### PR TITLE
feat(standups): Standup AI Summaries

### PR DESCRIPTION
**Pull request by jmtaber129**
# Description

Fixes https://github.com/ParabolInc/parabol/issues/8568

Implement AI Summaries for standups. Instead of prompting the model for an _overall_ summary, prompt the model to identify up to 5 themes found in responses, and provide brief summaries for those.

Reuse UI components for simplicity.

## Demo
This is the summary from several responses in our own July 17th Friday ship (copy-pasted to a local standup meeting).
Screen Shot 2023-07-26 at 9 25 09 AM

A different summary in slack: 
Screen Shot 2023-07-26 at 9 59 22 AM

## Testing scenarios
- [ ] No changes for orgs without the feature flag.
- [ ] Fill out a standup with test data (e.g. dummy standups or copied from a Friday ship)
- [ ] End the standup, and confirm the meeting immediately navigates to the summary, with the AI component loading
- [ ] Confirm the AI component displays a reasonable summary after a few seconds
- [ ] Confirm the email summary includes an AI summary
- [ ] For a meeting with 3+ responses, confirm that the `qualAIMeetingCount` on the `Team` object in PG is incremented.

## Final checklist

- [X] I checked the [code review guidelines](../docs/codeReview.md)
- [X] I have added [Metrics Representative](../docs/codeReview.md#metrics-representative) as reviewer(s) if my PR invovles metrics/data/analytics related changes
- [X] I have performed a self-review of my code, the same way I'd do it for any other team member
- [X] I have tested all cases I listed in the testing scenarios and I haven't found any issues or regressions
- [X] Whenever I took a non-obvious choice I added a comment explaining why I did it this way
- [X] I added the label `One Review Required` if the PR introduces only minor changes, does not contain any architectural changes or does not introduce any new patterns and I think one review is sufficient'
- [X] PR title is human readable and could be used in changelog

### CODE REVIEW

**Comment by nickoferrall** [comment]
Will review this tomorrow morning with fresh eyes!

File: packages/server/utils/OpenAIServerManager.ts
Comment (nickoferrall) [inline packages/server/utils/OpenAIServerManager.ts:37]: -1 davinci is the older model that Bruce is currently replacing with gpt-3.5-turbo here https://github.com/ParabolInc/parabol/pull/8227. That model is faster & cheaper, so we should probably start with it here or we'll need to replace it soon anyway.

File: packages/server/graphql/mutations/helpers/updateQualAIMeetingsCount.ts
Comment (nickoferrall) [inline packages/server/graphql/mutations/helpers/updateQualAIMeetingsCount.ts:21]: +1 The L&E mission is to increase Standups usage, so I wonder whether it'd be worth keeping the AI Summaries free for all users in Standups until you're happy with usage? 

The current implementation might help Self-Serve with conversions, so I won't advocate for my change too much 😄  but just an idea.

If you do want to keep it free, you'd need to remove the `canAccessAISummary` check, which has `return qualAIMeetingsCount < Threshold.MAX_QUAL_AI_MEETINGS`.

File: packages/client/modules/email/components/SummaryEmail/MeetingSummaryEmail/WholeMeetingSummaryResult.tsx
Comment (nickoferrall) [inline packages/client/modules/email/components/SummaryEmail/MeetingSummaryEmail/WholeMeetingSummaryResult.tsx:35]: -1 this could work for standups, but I don't think the Retro summary should be left aligned. 

Here's the left-aligned summary:

Screenshot 2023-07-28 at 11 36 06

And here's the summary without that change:
Screenshot 2023-07-28 at 11 36 17

File: packages/server/utils/OpenAIServerManager.ts
Comment (nickoferrall) [inline packages/server/utils/OpenAIServerManager.ts:28]: +1 I like this format, but it'd be great if the title on the client is bold rather than having an asterisk. 

When there are lots of speech marks, it can be difficult to read. 

Screenshot 2023-07-28 at 11 00 23

**Review by nickoferrall — CHANGES_REQUESTED** [review]
Looking good to me! I don't think this is a `One Review Required` PR though as there's a fair amount of new code and it's touching files related to the summary and ending a standup meeting

File: packages/client/modules/email/components/SummaryEmail/MeetingSummaryEmail/WholeMeetingSummaryResult.tsx
Comment (jmtaber129) [reply packages/client/modules/email/components/SummaryEmail/MeetingSummaryEmail/WholeMeetingSummaryResult.tsx:35]: I'm not sure I agree with this - center-alignment is fine for 2, _maybe_ 3 lines, but any more than that and it gets hard to read. Looking at our internal retros over the past couple months, most of our retro summaries are 3-4 lines. Examples:
* 🔒 https://action.parabol.co/new-summary/nCI2Ten1hm
* 🔒 https://action.parabol.co/new-summary/k0TSLF3pkI
* 🔒 https://action.parabol.co/new-summary/nnSHTZuMPC
* 🔒 https://action.parabol.co/new-summary/oh12yvDQ2Y

@enriquesanchez @acressall what do y'all think?

Comment (jmtaber129) [reply packages/client/modules/email/components/SummaryEmail/MeetingSummaryEmail/WholeMeetingSummaryResult.tsx:35]: I'm not sure I agree with this - center-alignment is fine for 2, _maybe_ 3 lines, but any more than that and it gets hard to read. Looking at our internal retros over the past couple months, most of our retro summaries are 3-4 lines. Examples:
* 🔒 https://action.parabol.co/new-summary/nCI2Ten1hm
* 🔒 https://action.parabol.co/new-summary/k0TSLF3pkI
* 🔒 https://action.parabol.co/new-summary/nnSHTZuMPC
* 🔒 https://action.parabol.co/new-summary/oh12yvDQ2Y

@enriquesanchez @acressall what do y'all think?

File: packages/server/graphql/mutations/helpers/updateQualAIMeetingsCount.ts
Comment (jmtaber129) [reply packages/server/graphql/mutations/helpers/updateQualAIMeetingsCount.ts:21]: L&E's focus this trimester is actually on signups, so I don't see a clear reason to make this one way or the other w.r.t. L&E's goals.

However, this makes me think that this would interfere with the trial-like experience for retro summaries. I.e. if a team holds daily recurring standups, they'd exceed the `MAX_QUAL_AI_MEETINGS` threshold within a few days.

I'll change this so that standup summaries are free for now.

File: packages/server/utils/OpenAIServerManager.ts
Comment (jmtaber129) [reply packages/server/utils/OpenAIServerManager.ts:28]: Oooo this is a really good idea! However, this would require some additional complexity when rendering the summary (we could just ask chatGPT for an HTML output and then do `dangerouslySetInnerHTML`, but that could open us to XSS).

I've created https://github.com/ParabolInc/parabol/issues/8604 to track this enhancement

File: packages/server/utils/OpenAIServerManager.ts
Comment (jmtaber129) [reply packages/server/utils/OpenAIServerManager.ts]: Fixed!

File: packages/client/modules/email/components/SummaryEmail/MeetingSummaryEmail/WholeMeetingSummaryResult.tsx
Comment (enriquesanchez) [reply packages/client/modules/email/components/SummaryEmail/MeetingSummaryEmail/WholeMeetingSummaryResult.tsx:35]: I agree that for longer texts left alignment is preferred for readability. We seem to be using both alignments in the same summary email (meeting summary is center aligned, topic summary is left aligned) and should stick to only one.

File: packages/server/utils/OpenAIServerManager.ts
Comment (acressall) [reply packages/server/utils/OpenAIServerManager.ts:28]: Might be a stretch, but could we use markdown so `* ` gets converted to a bullet point? I'm finding it difficult to read with all the asterisks and quotes. If not, maybe a dash would be easier to interpret than an asterisk?

Comment (jmtaber129) [reply packages/server/utils/OpenAIServerManager.ts]: @acressall absolutely! That's what https://github.com/ParabolInc/parabol/issues/8604 is intended to track - it's not a simple fix so I'd prefer to keep it out of this PR, but it would be great to prompt openAI for markdown or similar.

I went ahead and switched the asterisks to dashes per your recommendation, since it's definitely a bit cleaner and easier to read.

**Review by nickoferrall — APPROVED** [review]

File: packages/server/utils/OpenAIServerManager.ts
Comment (jmtaber129) [reply packages/server/utils/OpenAIServerManager.ts]: Oooo good catch! This was from an earlier iteration when I was trying out passing responses as HTML (it made the prompt too big FWIW).

Renamed to `plaintextResponses`

File: packages/client/modules/email/components/SummaryEmail/MeetingSummaryEmail/WholeMeetingSummary.tsx
Comment (mattkrick) [inline packages/client/modules/email/components/SummaryEmail/MeetingSummaryEmail/WholeMeetingSummary.tsx:59]: +1 i could be wrong here, but it looks like summary could be null if that fetch fails (e.g. openAI is down for a bit). Is there a way we could distinguish between an error state vs. a perpetual loading state?

File: packages/server/graphql/mutations/helpers/canAccessAISummary.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/helpers/canAccessAISummary.ts:18]: -1 this business logic is difficult to reason through.

I see `canAccessAISummary`, `isAISummaryAllowed` and then in another file i see 

```ts
const isAISummaryAccessible =
    (await isAISummaryAllowed(team, facilitator.featureFlags, dataLoader)) &&
    organization.featureFlags?.includes('standupAISummary')
```

can we reduce this to a single function that handles permissions for all `actor, action, resource` cases? Then when we start dropping feature flags it'll be as easy as deleting 1 line.

File: packages/server/graphql/mutations/helpers/safeEndTeamPrompt.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/helpers/safeEndTeamPrompt.ts:29]: -1 why `nonAtomic`?

**Review by mattkrick — CHANGES_REQUESTED** [review]
all minor stuff, excited to see this land!

File: packages/client/modules/email/components/SummaryEmail/MeetingSummaryEmail/WholeMeetingSummary.tsx
Comment (jmtaber129) [reply packages/client/modules/email/components/SummaryEmail/MeetingSummaryEmail/WholeMeetingSummary.tsx:59]: This is accurate, but this is just the current state of this component (not something new in this PR), so I'd prefer to tackle that separately.

I've created https://github.com/ParabolInc/parabol/issues/8615 to track this

File: packages/server/graphql/mutations/helpers/safeEndTeamPrompt.ts
Comment (jmtaber129) [reply packages/server/graphql/mutations/helpers/safeEndTeamPrompt.ts]: Copy-paste 🤦 removed

File: packages/server/graphql/mutations/helpers/canAccessAISummary.ts
Comment (jmtaber129) [reply packages/server/graphql/mutations/helpers/canAccessAISummary.ts]: Refactored this a bit

**Comment by mattkrick** [comment]
lgtm, couple merge conflicts, but after that feel free to merge

**Review by mattkrick — APPROVED** [review]

**Pull request merged**
